v2/crowdnotifier.ts

```diff
@@ -0,0 +1,220 @@
+    info: string;
+    nonce1: Uint8Array;
+    nonce2: Uint8Array;
+}
+
+export interface IEntProof {
+    nonce1: Uint8Array;
+    nonce2: Uint8Array;
+}
+
+export interface ILocationData {
+    ent: mcl.G2;
+    pEnt: IEntProof;
+    mtr: IMasterTrace;
+}
+
+export interface IPreTrace {
+    id: Uint8Array;
+    mskhEnc: Uint8Array; // In the paper ctxtHA
+    pskidl: mcl.G1;
+}
+
+export interface ITraceProof {
+    mpk: mcl.G2;
+    nonce1: Uint8Array;
+    nonce2: Uint8Array;
+}
+
+export interface ITrace {
+    id: Uint8Array;
+    skid: mcl.G1;
+}
+
+
+/**
+ * Implements the cryptographic protocol in section 4.2 of the CrowdNotifier whitepaper.
+ */
+export class CrowdNotifierPrimitives {
+
+    /**
+     * Generate key pair for health authority.
+     */
+    static setupHA(): IKeyPair {
+        return crypto_box_keypair();
+    }
+
+
+    /**
+     * Generate an identifier.
+     * @param info public information
+     * @param cnt counter
+     * @param nonce1 a nonce
+     * @param nonce2 an other nonce
+     * @returns an identifier
+     */
+    static genId(info: string, cnt: number, nonce1: Uint8Array, nonce2: Uint8Array): Uint8Array {
+
+        const hash1 = crypto_hash_sha256(
+            IBEIdInternal1.encode(
+                IBEIdInternal1.create({info, nonce: nonce1})
+            ).finish()
+        );
+
+        //id = H(H(info || nonce1) || cnt || nonce2)
+
+        const id = crypto_hash_sha256(
+            IBEIdInternal2.encode(
+                IBEIdInternal2.create({hash: hash1, cnt, nonce: nonce2})
+            ).finish()
+        );
+
+        return id;
+    }
+
+
+    /**
+     * Creates the data necessary for a location owner.
+     * @param info public info of the location, e.g., name:location:room
+     * @param pkha public key of the health authority
+     * @return the LocationData to be printed in multiple QR codes
+     */
+    static genCode(info: string, pkha: Uint8Array): ILocationData {
+        const [mpkh, mskh] = IBEPrimitives.keyGen();
+        const [mpkv, mskv] = IBEPrimitives.keyGen();
+        const mpk = mcl.add(mpkh, mpkv);
+        const nonce1 = randombytes_buf(NONCE_LENGTH);
+        const nonce2 = randombytes_buf(NONCE_LENGTH);
+
+        // In the paper ctxtHA
+        const mskhEnc = crypto_box_seal(mskh.serialize(), pkha);
+
+        const pEnt = {nonce1, nonce2};
+        const mtr = {mpk, mskv, mskhEnc, info, nonce1, nonce2};
+
+        return {ent: mpk, pEnt, mtr}
+    }
+
+
+    /**
+     * Creates a private record for the user to store in her phone.
+     * @param ent as presented in the QRentry
+     * @param pEnt as presented in the QRentry
+     * @param info as presented in the QRentry
+     * @param cnt count of hours since UNIX epoch
+     * @param aux free data
+     * @return user record to be stored
+     */
+    static scan(ent: mcl.G2, pEnt: IEntProof, info: string, cnt: number, aux: string): IEncryptedData {
+        const aux_u = new TextEncoder().encode(aux);
```

Comment:
Once `aux` is of type `Uint8Array`, then this will go away. We don't want to handle text encoding in the crypto part - that is a problem of it's own and should be handled in the higher layers...